sender.go

```diff
@@ -297,12 +296,44 @@ Loop:
 			outgoingTransfers = s.transfers
 		}
 
+		if len(outgoingDisps) > 0 && len(outgoingDisp) == 0 {
+			// queue up the next outgoing frame and remove it from the slice
+			outgoingDisp <- outgoingDisps[0]
+			outgoingDisps = outgoingDisps[1:]
+		}
+
 		select {
+		case dr := <-outgoingDisp:
+			debug.Log(3, "TX (sender): %s", dr)
+
+			// Ensure the session mux is not blocked
+			for {
+				select {
+				case s.l.session.tx <- dr:
+					continue Loop
+				case fr := <-s.l.rx:
+					var disp *frames.PerformDisposition
+					disp, s.l.err = s.muxHandleFrame(fr)
+					if s.l.err != nil {
+						return
+					} else if disp != nil {
+						outgoingDisps = append(outgoingDisps, disp)
+					}
+				case <-s.l.close:
+					continue Loop
+				case <-s.l.session.done:
+					continue Loop
+				}
+			}
+
 		// received frame
 		case fr := <-s.l.rx:
-			s.l.err = s.muxHandleFrame(fr)
+			var disp *frames.PerformDisposition
+			disp, s.l.err = s.muxHandleFrame(fr)
```

Comment:
There's some duplication here that might be a bit error prone - maybe we can factor it into a local function and call it at 315, 331, and 356?

```go
handleFrame := func(fr *amqp.Frame) {
  var disp *frames.PerformDisposition
  disp, s.l.err = s.muxHandleFrame(fr)
  if s.l.err != nil {
    return
  } else if disp != nil {
    outgoingDisps = append(outgoingDisps, disp)
  }
}
```

Reply:
Good idea, fixed.